node/src/components/fetcher/trie_fetcher.rs

```diff
@@ -0,0 +1,342 @@
+}
+
+#[derive(DataSize, Debug, From)]
+pub(crate) enum Event<I>
+where
+    I: Debug + Eq,
+{
+    #[from]
+    Request(TrieFetcherRequest<I>),
+    TrieOrChunkFetched {
+        id: TrieOrChunkedDataId,
+        fetch_result: FetchResult<TrieOrChunkedData, I>,
+    },
+}
+
+impl<I> fmt::Display for Event<I>
+where
+    I: Debug + Eq,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Event::Request(_) => write!(f, "trie fetcher request"),
+            Event::TrieOrChunkFetched { id, .. } => write!(f, "trie or chunk {} fetched", id),
+        }
+    }
+}
+
+fn into_response<I>(trie: Box<Trie<Key, StoredValue>>, sender: Option<I>) -> TrieFetcherResult<I>
+where
+    I: Debug + Eq,
+{
+    match sender {
+        Some(peer) => Ok(FetchedData::FromPeer { item: trie, peer }),
+        None => Ok(FetchedData::FromStorage { item: trie }),
+    }
+}
+
+impl<I> TrieFetcher<I>
+where
+    I: Debug + Clone + Send + Eq + 'static,
+{
+    pub(crate) fn new() -> Self {
+        TrieFetcher {
+            partial_chunks: Default::default(),
+        }
+    }
+
+    fn consume_trie_or_chunk<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        sender: Option<I>,
+        trie_or_chunk: TrieOrChunkedData,
+    ) -> Effects<Event<I>>
+    where
+        REv: ReactorEventT<TrieOrChunkedData> + From<FetcherRequest<I, TrieOrChunkedData>>,
+    {
+        let TrieOrChunkedDataId(_index, hash) = trie_or_chunk.id();
+        match trie_or_chunk {
+            TrieOrChunkedData::Trie(trie) => match self.partial_chunks.remove(&hash) {
+                None => {
+                    error!(%hash, "fetched a trie we didn't request!");
+                    Effects::new()
+                }
+                Some(partial_chunks) => {
+                    debug!(%hash, "got a full trie");
+                    partial_chunks
+                        .responder
+                        .respond(into_response(trie, sender))
+                        .ignore()
+                }
+            },
+            TrieOrChunkedData::ChunkWithProof(chunk) => {
+                self.consume_chunk(effect_builder, sender, chunk)
+            }
+        }
+    }
+
+    fn consume_chunk<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        sender: Option<I>,
+        chunk: ChunkWithProof,
+    ) -> Effects<Event<I>>
+    where
+        REv: ReactorEventT<TrieOrChunkedData> + From<FetcherRequest<I, TrieOrChunkedData>>,
+    {
+        if !chunk.verify() {
+            match sender {
+                None => {
+                    error!(?chunk, "got an invalid chunk from storage");
+                    return Effects::new();
+                }
+                Some(sender) => {
+                    warn!(?sender, ?chunk, "got an invalid chunk from sender");
+                    // TODO: would be good to re-request from someone else instead of the same
+                    // node...
```

Comment:
Indeed. :smile: 
Also, maybe we should disconnect from that peer.

Reply:
Oops :sweat_smile: It's a leftover from the initial implementation. I've added a way of trying multiple peers since then, but it seems that I forgot to use it in this place :sweat_smile: 

And disconnecting seems like a good idea to me :+1: